### PR TITLE
fix: check the pointer is valid

### LINKED ISSUE #851: Segfault when printing a list of `DataFrame`s created with `future`
When using `lapply()`, all good:

``` r
library(future.apply)
#> Loading required package: future
library(polars)
plan(multisession)

do_stuff <- function(n){
  as_polars_df(data.frame(x = 1:3))
}

n <- 2
out <- lapply(1:n, do_stuff)
# out <- future_lapply(1:n, do_stuff)
out
#> [[1]]
#> shape: (3, 1)
#> ┌─────┐
#> │ x   │
#> │ --- │
#> │ i32 │
#> ╞═════╡
#> │ 1   │
#> │ 2   │
#> │ 3   │
#> └─────┘
#> 
#> [[2]]
#> shape: (3, 1)
#> ┌─────┐
#> │ x   │
#> │ --- │
#> │ i32 │
#> ╞═════╡
#> │ 1   │
#> │ 2   │
#> │ 3   │
#> └─────┘
```

When using `future_lapply()`, crash:

This reprex appears to crash R.
See standard output and standard error for more details.

#### Standard output and error

``` sh

 *** caught segfault ***
address 0x8, cause 'memory not mapped'

Traceback:
 1: .pr$DataFrame$print(self)
 2: x$print()
 3: print.RPolarsDataFrame(x)
 4: (function (x, ...) UseMethod("print"))(x)
 5: print.default(x)
 6: print(x)
 7: normal_print(x)
 8: knit_print.default(x, ...)
 9: knit_print(x, ...)
10: withVisible(knit_print(x, ...))
11: fun(x, options = options)
12: value_fun(ev$value, ev$visible)
13: withVisible(value_fun(ev$value, ev$visible))
14: withCallingHandlers(withVisible(value_fun(ev$value, ev$visible)),     warning = wHandler, error = eHandler, message = mHandler)
15: doTryCatch(return(expr), name, parentenv, handler)
16: tryCatchOne(expr, names, parentenv, handlers[[1L]])
17: tryCatchList(expr, classes, parentenv, handlers)
18: tryCatch(expr, error = function(e) {    call <- conditionCall(e)    if (!is.null(call)) {        if (identical(call[[1L]], quote(doTryCatch)))             call <- sys.call(-4L)        dcall <- deparse(call, nlines = 1L)        prefix <- paste("Error in", dcall, ": ")        LONG <- 75L        sm <- strsplit(conditionMessage(e), "\n")[[1L]]        w <- 14L + nchar(dcall, type = "w") + nchar(sm[1L], type = "w")        if (is.na(w))             w <- 14L + nchar(dcall, type = "b") + nchar(sm[1L],                 type = "b")        if (w > LONG)             prefix <- paste0(prefix, "\n  ")    }    else prefix <- "Error : "    msg <- paste0(prefix, conditionMessage(e), "\n")    .Internal(seterrmessage(msg[1L]))    if (!silent && isTRUE(getOption("show.error.messages"))) {        cat(msg, file = outFile)        .Internal(printDeferredWarnings())    }    invisible(structure(msg, class = "try-error", condition = e))})
19: try(f, silent = TRUE)
20: handle(pv <- withCallingHandlers(withVisible(value_fun(ev$value,     ev$visible)), warning = wHandler, error = eHandler, message = mHandler))
21: evaluate_call(expr, parsed$src[[i]], envir = envir, enclos = enclos,     debug = debug, last = i == length(out), use_try = stop_on_error !=         2L, keep_warning = keep_warning, keep_message = keep_message,     log_echo = log_echo, log_warning = log_warning, output_handler = output_handler,     include_timing = include_timing)
22: evaluate::evaluate(...)
23: evaluate(code, envir = env, new_device = FALSE, keep_warning = if (is.numeric(options$warning)) TRUE else options$warning,     keep_message = if (is.numeric(options$message)) TRUE else options$message,     stop_on_error = if (is.numeric(options$error)) options$error else {        if (options$error && options$include)             0L        else 2L    }, output_handler = knit_handlers(options$render, options))
24: in_dir(input_dir(), expr)
25: in_input_dir(evaluate(code, envir = env, new_device = FALSE,     keep_warning = if (is.numeric(options$warning)) TRUE else options$warning,     keep_message = if (is.numeric(options$message)) TRUE else options$message,     stop_on_error = if (is.numeric(options$error)) options$error else {        if (options$error && options$include)             0L        else 2L    }, output_handler = knit_handlers(options$render, options)))
26: eng_r(options)
27: block_exec(params)
28: call_block(x)
29: process_group.block(group)
30: process_group(group)
31: withCallingHandlers(if (tangle) process_tangle(group) else process_group(group),     error = function(e) if (xfun::pkg_available("rlang", "1.0.0")) rlang::entrace(e))
32: withCallingHandlers(expr, error = function(e) {    loc = paste0(current_lines(), label, sprintf(" (%s)", knit_concord$get("infile")))    message(one_string(handler(e, loc)))})
33: handle_error(withCallingHandlers(if (tangle) process_tangle(group) else process_group(group),     error = function(e) if (xfun::pkg_available("rlang", "1.0.0")) rlang::entrace(e)),     function(e, loc) {        setwd(wd)        write_utf8(res, output %n% stdout())        paste0("\nQuitting from lines ", loc)    }, if (labels[i] != "") sprintf(" [%s]", labels[i]))
34: process_file(text, output)
35: knitr::knit(knit_input, knit_output, envir = envir, quiet = quiet)
36: rmarkdown::render(input, quiet = TRUE, envir = globalenv(), encoding = "UTF-8")
37: (function (input) {    rmarkdown::render(input, quiet = TRUE, envir = globalenv(),         encoding = "UTF-8")})(input = base::quote("keen-boa_reprex.R"))
38: (function (what, args, quote = FALSE, envir = parent.frame()) {    if (!is.list(args))         stop("second argument must be a list")    if (quote)         args <- lapply(args, enquote)    .Internal(do.call(what, args, envir))})(base::quote(function (input) {    rmarkdown::render(input, quiet = TRUE, envir = globalenv(),         encoding = "UTF-8")}), base::quote(list(input = "keen-boa_reprex.R")), envir = base::quote(<environment>),     quote = base::quote(TRUE))
39: do.call(do.call, c(readRDS("/tmp/RtmpAu8T1Y/callr-fun-29201683346ae"),     list(envir = .GlobalEnv, quote = TRUE)), envir = .GlobalEnv,     quote = TRUE)
40: saveRDS(do.call(do.call, c(readRDS("/tmp/RtmpAu8T1Y/callr-fun-29201683346ae"),     list(envir = .GlobalEnv, quote = TRUE)), envir = .GlobalEnv,     quote = TRUE), file = "/tmp/RtmpAu8T1Y/callr-res-2920161195e01",     compress = FALSE)
41: withCallingHandlers({    NULL    saveRDS(do.call(do.call, c(readRDS("/tmp/RtmpAu8T1Y/callr-fun-29201683346ae"),         list(envir = .GlobalEnv, quote = TRUE)), envir = .GlobalEnv,         quote = TRUE), file = "/tmp/RtmpAu8T1Y/callr-res-2920161195e01",         compress = FALSE)    flush(stdout())    flush(stderr())    NULL    invisible()}, error = function(e) {    {        callr_data <- as.environment("tools:callr")$`__callr_data__`        err <- callr_data$err        if (FALSE) {            assign(".Traceback", .traceback(4), envir = callr_data)            dump.frames("__callr_dump__")            assign(".Last.dump", .GlobalEnv$`__callr_dump__`,                 envir = callr_data)            rm("__callr_dump__", envir = .GlobalEnv)        }        e <- err$process_call(e)        e2 <- err$new_error("error in callr subprocess")        class(e2) <- c("callr_remote_error", class(e2))        e2 <- err$add_trace_back(e2)        cut <- which(e2$trace$scope == "global")[1]        if (!is.na(cut)) {            e2$trace <- e2$trace[-(1:cut), ]        }        saveRDS(list("error", e2, e), file = paste0("/tmp/RtmpAu8T1Y/callr-res-2920161195e01",             ".error"))    }}, interrupt = function(e) {    {        callr_data <- as.environment("tools:callr")$`__callr_data__`        err <- callr_data$err        if (FALSE) {            assign(".Traceback", .traceback(4), envir = callr_data)            dump.frames("__callr_dump__")            assign(".Last.dump", .GlobalEnv$`__callr_dump__`,                 envir = callr_data)            rm("__callr_dump__", envir = .GlobalEnv)        }        e <- err$process_call(e)        e2 <- err$new_error("error in callr subprocess")        class(e2) <- c("callr_remote_error", class(e2))        e2 <- err$add_trace_back(e2)        cut <- which(e2$trace$scope == "global")[1]        if (!is.na(cut)) {            e2$trace <- e2$trace[-(1:cut), ]        }        saveRDS(list("error", e2, e), file = paste0("/tmp/RtmpAu8T1Y/callr-res-2920161195e01",             ".error"))    }}, callr_message = function(e) {    try(signalCondition(e))})
42: doTryCatch(return(expr), name, parentenv, handler)
43: tryCatchOne(expr, names, parentenv, handlers[[1L]])
44: tryCatchList(expr, names[-nh], parentenv, handlers[-nh])
45: doTryCatch(return(expr), name, parentenv, handler)
46: tryCatchOne(tryCatchList(expr, names[-nh], parentenv, handlers[-nh]),     names[nh], parentenv, handlers[[nh]])
47: tryCatchList(expr, classes, parentenv, handlers)
48: tryCatch(withCallingHandlers({    NULL    saveRDS(do.call(do.call, c(readRDS("/tmp/RtmpAu8T1Y/callr-fun-29201683346ae"),         list(envir = .GlobalEnv, quote = TRUE)), envir = .GlobalEnv,         quote = TRUE), file = "/tmp/RtmpAu8T1Y/callr-res-2920161195e01",         compress = FALSE)    flush(stdout())    flush(stderr())    NULL    invisible()}, error = function(e) {    {        callr_data <- as.environment("tools:callr")$`__callr_data__`        err <- callr_data$err        if (FALSE) {            assign(".Traceback", .traceback(4), envir = callr_data)            dump.frames("__callr_dump__")            assign(".Last.dump", .GlobalEnv$`__callr_dump__`,                 envir = callr_data)            rm("__callr_dump__", envir = .GlobalEnv)        }        e <- err$process_call(e)        e2 <- err$new_error("error in callr subprocess")        class(e2) <- c("callr_remote_error", class(e2))        e2 <- err$add_trace_back(e2)        cut <- which(e2$trace$scope == "global")[1]        if (!is.na(cut)) {            e2$trace <- e2$trace[-(1:cut), ]        }        saveRDS(list("error", e2, e), file = paste0("/tmp/RtmpAu8T1Y/callr-res-2920161195e01",             ".error"))    }}, interrupt = function(e) {    {        callr_data <- as.environment("tools:callr")$`__callr_data__`        err <- callr_data$err        if (FALSE) {            assign(".Traceback", .traceback(4), envir = callr_data)            dump.frames("__callr_dump__")            assign(".Last.dump", .GlobalEnv$`__callr_dump__`,                 envir = callr_data)            rm("__callr_dump__", envir = .GlobalEnv)        }        e <- err$process_call(e)        e2 <- err$new_error("error in callr subprocess")        class(e2) <- c("callr_remote_error", class(e2))        e2 <- err$add_trace_back(e2)        cut <- which(e2$trace$scope == "global")[1]        if (!is.na(cut)) {            e2$trace <- e2$trace[-(1:cut), ]        }        saveRDS(list("error", e2, e), file = paste0("/tmp/RtmpAu8T1Y/callr-res-2920161195e01",             ".error"))    }}, callr_message = function(e) {    try(signalCondition(e))}), error = function(e) {    NULL    if (TRUE) {        try(stop(e))    }    else {        invisible()    }}, interrupt = function(e) {    NULL    if (TRUE) {        e    }    else {        invisible()    }})
An irrecoverable exception occurred. R is aborting now ...
```

### PR DESCRIPTION
**Pull request by eitsupi**
Fix #851

I feel this ideally be implemented in extendr (or savvy etc.), but I will implement it here anyway.
cc @yutannihilation @CGMossa @JosiahParry

### CODE REVIEW

**Comment by eitsupi** [comment]
It seems to have little effect on performance.

Installed from 12d4bc2

``` r
library(polars)

countries = c(
  "France", "Germany", "United Kingdom", "Japan", "Columbia",
  "South Korea", "Vietnam", "South Africa", "Senegal", "Iran"
)

set.seed(123)
df_test = pl$DataFrame(
  country = sample(countries, 1e7, TRUE),
  x = sample(1:100, 1e7, TRUE),
  y = sample(1:1000, 1e7, TRUE)
)

lf_test = df_test$lazy()

lazy_query = lf_test$
  sort(pl$col("country"))$
  filter(
  pl$col("country")$is_in(pl$lit(c("United Kingdom", "Japan", "Vietnam")))
)

bench::mark(
  eager = df_test$
    sort(pl$col("country"))$
    filter(
      pl$col("country")$is_in(pl$lit(c("United Kingdom", "Japan", "Vietnam")))
    ),
  lazy = lazy_query$collect(),
  iterations = 10
)
#> # A tibble: 2 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 eager         911ms    1.07s     0.934   14.56KB        0
#> 2 lazy          488ms 593.78ms     1.71     1.94KB        0
```

<sup>Created on 2024-03-01 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

installed from main (bec40c6929cdd6ff6ae5bb56da0618bb4bc83d92)

``` r
library(polars)

countries = c(
  "France", "Germany", "United Kingdom", "Japan", "Columbia",
  "South Korea", "Vietnam", "South Africa", "Senegal", "Iran"
)

set.seed(123)
df_test = pl$DataFrame(
  country = sample(countries, 1e7, TRUE),
  x = sample(1:100, 1e7, TRUE),
  y = sample(1:1000, 1e7, TRUE)
)

lf_test = df_test$lazy()

lazy_query = lf_test$
  sort(pl$col("country"))$
  filter(
  pl$col("country")$is_in(pl$lit(c("United Kingdom", "Japan", "Vietnam")))
)

bench::mark(
  eager = df_test$
    sort(pl$col("country"))$
    filter(
      pl$col("country")$is_in(pl$lit(c("United Kingdom", "Japan", "Vietnam")))
    ),
  lazy = lazy_query$collect(),
  iterations = 10
)
#> # A tibble: 2 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 eager         978ms    1.05s     0.906   12.46KB        0
#> 2 lazy          532ms 574.97ms     1.72     1.94KB        0
```

<sup>Created on 2024-03-01 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

File: R/polars-package.R [hunk 1]
```diff
@@ -2,7 +2,9 @@
 "_PACKAGE"
 
 ## usethis namespace: start
-#' @importFrom utils globalVariables head tail download.file capture.output str
+#' @importFrom stats na.omit median
```
Comment (eitsupi) [inline R/polars-package.R:5]: I do not know if this is necessary.

Comment (etiennebacher) [reply R/polars-package.R:5]: We add methods for those so I guess they should be imported

**Comment by CGMossa** [comment]
Hello! Allow me to give my two cents on this.

- A null in R, `NULL`, is an `SEXP` with type `NILSXP`. This means that when you set a variable to `NULL`, you're changing the type to a "disposable" one.
- A null-check for an external-ptr consists of checking whether the externalptr attributed to an `SEXP` is in fact a NULL in terms of C, meaning that the value of the pointer is 0. A `NULL` in R's C-API is a `R_NilValue`, which is not a C NULL. 
- Do you return `&self`/`&mut self` in any of the `#[extendr]`-impl blocks here? I'm about to check that myself.
-

**Review by etiennebacher — CHANGES_REQUESTED** [review]
~Thanks, I confirm the test doesn't segfault locally.~ I'm gonna let @CGMossa and others discuss about the best method here, I don't have enough knowledge for that.

Can you bump NEWS?

File: R/polars-package.R (hunk 1 above)
Comment (eitsupi) [reply R/polars-package.R:5]: Other functions like mean are not imported but seem to work.

Comment (etiennebacher) [reply R/polars-package.R:5]: Yes but `mean()` is in `base` and not in `stats` (and I guess the other functions as well)

Comment (eitsupi) [reply R/polars-package.R:5]: Oh, I see. Thanks.

**Comment by CGMossa** [comment]
I've been experimenting and fixing how we do externalptr in `extendr` on https://github.com/cgmossa/extendr/. It should become more robust soon. I tend to focus on the wrong thing, so allow me to ask you this: Where do you believe the null-checking should occur, and what should it do?

**Comment by yutannihilation** [comment]
Here's my take. Hope this helps.

https://github.com/yutannihilation/savvy/pull/95/files#diff-e2092a335ec465830416f879103b89ea75be8e414065e9e934cd82697ea63ea6

**Review by etiennebacher — APPROVED** [review]

**Pull request merged**